libkbfs/tlf_handle.go

```diff
@@ -931,40 +919,78 @@ func ParseTlfHandle(
 	if len(extensionSuffix) != 0 {
 		extensions, err = ParseTlfHandleExtensionSuffix(extensionSuffix)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 	}
 
 	h, err := makeTlfHandleHelper(ctx, public, writers, readers, extensions)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	if !public {
 		currentUsername, currentUID, err := kbpki.GetCurrentUserInfo(ctx)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 
 		if !h.IsReader(currentUID) {
-			return nil, ReadAccessError{currentUsername, h.GetCanonicalName(), public}
+			return nil, nil, ReadAccessError{currentUsername, h.GetCanonicalName(), public}
 		}
 	}
 
-	if string(h.GetCanonicalName()) == name {
-		// Name is already canonical (i.e., all usernames and
-		// no assertions) so we can delay the identify until
-		// the node is actually used.
-		return h, nil
-	}
+	switch identifyBehavior {
+	case keybase1.TLFIdentifyBehavior_DEFAULT_KBFS:
+		if string(h.GetCanonicalName()) == name {
+			// Name is already canonical (i.e., all usernames and
+			// no assertions) so we can delay the identify until
+			// the node is actually used.
+			return h, nil, nil
+		}
 
-	// Otherwise, identify before returning the canonical name.
-	err = identifyHandle(ctx, kbpki, kbpki, h)
-	if err != nil {
-		return nil, err
+		// Otherwise, identify before returning the canonical name.
+		err = identifyHandle(ctx, kbpki, kbpki, h)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return nil, nil, TlfNameNotCanonical{name, string(h.GetCanonicalName())}
+	case keybase1.TLFIdentifyBehavior_CHAT:
+		breaks, err := identifyHandleWithIdentifyBehavior(ctx, kbpki, kbpki, h,
+			keybase1.TLFIdentifyBehavior_CHAT)
+		if err != nil {
+			return nil, breaks, err
+		}
+
+		if string(h.GetCanonicalName()) != name {
```

Comment:
Why does this come after the identifies in this case?  I guess you're trying to force identifies to happen during parsing, to avoid plumbing `behavior` all the way through `KBFSOps`?  I don't really think that strategy is going to work, since in the case where we're initializing a new folder that didn't exist before, KBFS will do another identify with the default behavior, which will break if there are broken tracker statements.

I think it would be better to still short-circuit the identifies here if the name is canonical, and instead plumb `behavior` into `KBFSOpsStandard.getOrInitializeNewMDMaster`.  Then, you'd have to add an explicit call to identify somewhere if a) the TLF already existed on the servers, and b) there isn't yet a `folderBranchOps` with `identifyDone=true`.  And when creating a new folder, you'd have to plumb the behavior through to `folderBranchOps.SetInitialHeadToNew`.  I can't see a way around this.

If it's too painful to plumb the behavior forward and the breaks backwards through all those codepaths, it probably makes sense to use the context strategy like we discussed on the phone call.